packages/actor-optimize-link-traversal-filter-tree-links/lib/ActorOptimizeLinkTraversalFilterTreeLinks.ts

```diff
@@ -0,0 +1,167 @@
+import { BindingsFactory } from '@comunica/bindings-factory';
+import type {
+  IActorOptimizeLinkTraversalArgs,
+  IActionOptimizeLinkTraversal,
+  IActorOptimizeLinkTraversalOutput,
+} from '@comunica/bus-optimize-link-traversal';
+import { ActorOptimizeLinkTraversal } from '@comunica/bus-optimize-link-traversal';
+import { KeysInitQuery } from '@comunica/context-entries';
+import type { IActorTest } from '@comunica/core';
+import type { Bindings } from '@comunica/types';
+import type { IRelation } from '@comunica/types-link-traversal';
+import type * as RDF from 'rdf-js';
+import { stringToTerm } from 'rdf-string';
+import { Algebra } from 'sparqlalgebrajs';
+import { AsyncEvaluator } from 'sparqlee';
+
+/**
+ * A comunica Link traversal optimizer that filter link of document
+ * following the [TREE specification](https://treecg.github.io/specification/)
+ */
+
+export class ActorOptimizeLinkTraversalFilterTreeLinks extends ActorOptimizeLinkTraversal {
+  public constructor(args: IActorOptimizeLinkTraversalArgs) {
+    super(args);
+  }
+
+  public async test(action: IActionOptimizeLinkTraversal): Promise<IActorTest> {
+    const query: Algebra.Operation = action.context.get(KeysInitQuery.query)!;
+    const relations: IRelation[] = (() => {
+      if (typeof action.treeMetadata !== 'undefined') {
+        return typeof action.treeMetadata.relation !== 'undefined' ? action.treeMetadata.relation : [];
+      }
+      return [];
+    })();
+    const filterExist: boolean = this.doesNodeExist(query, Algebra.types.FILTER);
+    return filterExist && relations.length > 0 ?
+      Promise.resolve(true) :
+      Promise.reject(new Error(
+        'the action must contain TREE relation and the query must contain at least a filter',
+      ));
+  }
+
+  public async run(action: IActionOptimizeLinkTraversal): Promise<IActorOptimizeLinkTraversalOutput> {
+    const filterMap: Map<string, boolean> = new Map();
+    const filterOperation: Algebra.Expression = JSON.parse(JSON.stringify(action.context.get(KeysInitQuery.query)))
```

Comment:
because of the shallow copy it modify the filter expression

Reply:
This is an unsafe way of copying objects, as it will not include methods.
There should be a helper function for copying operations in sparqlalgebrajs.